Sources/Sentry/SentrySessionReplay.m

```diff
@@ -0,0 +1,250 @@
+#import "SentryDependencyContainer.h"
+#import "SentryFileManager.h"
+#import "SentryHub+Private.h"
+#import "SentryLog.h"
+#import "SentryReplayEvent.h"
+#import "SentryReplayRecording.h"
+#import "SentrySDK+Private.h"
+#import "SentrySwift.h"
+
+#if SENTRY_HAS_UIKIT && !TARGET_OS_VISION
+static NSString *SENTRY_REPLAY_FOLDER = @"replay";
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation SentrySessionReplay {
+    UIView *_rootView;
+    BOOL _processingScreenshot;
+    CADisplayLink *_displayLink;
+    NSDate *_lastScreenShot;
+    NSDate *_videoSegmentStart;
+    NSURL *_urlToCache;
+    NSDate *_sessionStart;
+    SentryReplayOptions *_replayOptions;
+    SentryOnDemandReplay *_replayMaker;
+    SentryId *sessionReplayId;
+    NSMutableArray<UIImage *> *imageCollection;
+    int _currentSegmentId;
+    BOOL _isFullSession;
+    SentryCurrentDateProvider *_dateProvider;
+}
+
+- (instancetype)initWithSettings:(SentryReplayOptions *)replayOptions
+                    dateProvider:(SentryCurrentDateProvider *)dateProvider
+{
+    if (self = [super init]) {
+        _replayOptions = replayOptions;
+        _dateProvider = dateProvider;
+    }
+    return self;
+}
+
+- (void)start:(UIView *)rootView fullSession:(BOOL)full
+{
+    if (rootView == nil) {
+        SENTRY_LOG_DEBUG(@"rootView cannot be nil. Session replay will not be recorded.");
+        return;
+    }
+
+    @synchronized(self) {
+        if (_displayLink == nil) {
+            _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(newFrame:)];
+            [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
+        } else {
+            // Session display is already running.
+            return;
+        }
+
+        _rootView = rootView;
+        _lastScreenShot = [[NSDate alloc] init];
+        _videoSegmentStart = nil;
+        _sessionStart = _lastScreenShot;
+        _currentSegmentId = 0;
+
+        NSURL *docs = [NSURL
+            fileURLWithPath:[SentryDependencyContainer.sharedInstance.fileManager sentryPath]];
+        docs = [docs URLByAppendingPathComponent:SENTRY_REPLAY_FOLDER];
+
+        NSString *currentSession = [NSUUID UUID].UUIDString;
+        _urlToCache = [docs URLByAppendingPathComponent:currentSession];
+
+        if (![NSFileManager.defaultManager fileExistsAtPath:_urlToCache.path]) {
+            [NSFileManager.defaultManager createDirectoryAtURL:_urlToCache
+                                   withIntermediateDirectories:YES
+                                                    attributes:nil
+                                                         error:nil];
+        }
+
+        _replayMaker = [[SentryOnDemandReplay alloc] initWithOutputPath:_urlToCache.path];
+        _replayMaker.bitRate = _replayOptions.replayBitRate;
+        _replayMaker.cacheMaxSize = (NSInteger)(full ? _replayOptions.sessionSegmentDuration
+                                                     : _replayOptions.errorReplayDuration);
```

Comment:
I don't quite understand what's being done here. What is a segment? Is this a customer-facing option? We should explain this in detail in a headerdoc there.